### ISSUE TITLE
NoGradContext blocked by UnsafePointer parametric mutability

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2400]
## Description

The NoGradContext implementation is blocked by a Mojo language limitation regarding UnsafePointer with parametric mutability.

## Location

- File: `shared/autograd/tape.mojo`
- Line: 805

## TODO/FIXME Text

```
# TODO: NoGradContext requires UnsafePointer with parametric mutability
```

## Background

NoGradContext is meant to temporarily disable gradient computation (like PyTorch's `torch.no_grad()`):

```python
# PyTorch equivalent
with torch.no_grad():
    output = model(input)  # No gradients tracked
```

## Blocker

The implementation requires UnsafePointer with parametric mutability, which may not be fully supported in the current Mojo version.

## Acceptance Criteria

- [ ] Investigate current Mojo UnsafePointer capabilities
- [ ] Implement NoGradContext if possible with current API
- [ ] Or document workaround until Mojo supports this
- [ ] Context manager semantics (__enter__, __exit__)
- [ ] Thread-safety considerations
- [ ] Tests for gradient disabling behavior

## Workaround

Until this is implemented, users may need to manually manage gradient tracking flags.

## Notes

This may require waiting for Mojo language updates. Track Mojo changelog for UnsafePointer improvements.
